### ISSUE TITLE
code requests

### ISSUE THREAD

**Issue by junkangwu** [Linear95/CLUB#6]
Hi, Pengyu~
I'm interested in mutual information minimization and its application. Thanks a lot for your excellent work in this area.
As for the application of mutual information minimization,  I came across your work on ACL2020'Improving Disentangled Text Representation Learning with Information-Theoretic Guidance'  and I found it really fascinating!
Could you please share the code with me so that I can better understand the implementation?

Thank you!

**Comment by Linear95** [comment]
Thank you for being interested in our ACL 2020 paper. That work was conducted when I was doing an internship at NEC Labs, and is currently under a patent application process. I haven't got the right to publish the code due to the company's policy. I will publish the resources when I get permission. Thanks.

**Comment by junkangwu** [comment]
Thanks for your reply. By the way, May I ask you for some details on the mutual information minimization? 
1. To realize the mutual information minimization, I follow your model in 2020ICML-Club to calculate the upbound of mutual information. And I try the representation of the `club_sample` in the `mi_estimators.py` to minimize the mutual information between the h_1 and h_2 which are the hidden representations after the encoder process. However, I found it difficult to converge on my training process. Unfortunately, I cannot converge well with it in my own model, and even negative values often occur. Could you share with me some suggestions or experiences on the application about the `club_sample` estimator?
2. In your method of calculating the `mu and log var`, you implement it with two fully-connected neural networks. May I ask you about the choices on this estimation？Will the choice of this method influence my mutual information upper bound minimization?
Thanks a lot in advance!

**Comment by Linear95** [comment]
I just upload a toy example of MI minimization under Gaussian setups. Hope it could help you understand our MI minimization algorithm. For your questions:  

1. Yes, the structure of q(y|x) influence the performance significantly. The estimators in this repo are under the simplest cases. You may change the architecture of q(y|x) to fit your project. 

2. The training of MI minimization is somehow similar to the training of GAN. I think you can try different learning rates and different structures of q(y|x) for the MI estimator. If the estimation is negative, that is because your MI estimator is not well learned. You need to update the parameters of the MI estimator (by maximizing the loglikelihood) more frequently or enlarge the learning rate.

**Comment by junkangwu** [comment]
Thanks for your detailed sharing and suggestions. By the way, May I ask you about the MI estimator in your CLUB model's application in the `2020ACL` paper? Actually, I just desire to minimize the mutual information between the representations after GCN_1 and GCN_2. In other words, that is two same encoders but not sharing parameters. 
Thanks a lot!

**Comment by Linear95** [comment]
It is the same estimator I used in this repo. But you may need to change the hidden_size of CLUB depending on your representation dimension.

**Comment by junkangwu** [comment]
Thanks a lot for your reply. I will have a try!

**Comment by junkangwu** [comment]
I'm sorry to disturb you again. But I am still confused about the training process for the parameter updating. 
1. May I ask you about the order of parameters updating? Following the Algorithm1(MI Minimization with vCLUB), (a)when we are maximizing the log-likelihood of q(y|x),  only the parameters in `vCLUB` could be updated? And in the next process, (b) when we need to minimize the upper_bound loss, the parameters in `vCLUB` should be fixed, and only the parameters in the `encoder` module could be updated? Are there any errors in my statement or understanding? 
2. At the beginning of the training, I did not conduct free or frozen for any module parameters of the training process, and the result was log Likehod loss which decreased from 70 to more than -200 and could not be converged. After adding the parameter-free or frozen process, as described in 1.  At this point, the LLD loss decreases a little, and the mutual information upper bound loss becomes negative to more than -100. 

Perhaps I'm new to train some structure of GAN, I think I understand your theory but the results of the experiment are so strange that I'm confused and don't know how to go!
 
Thanks a lot in advance sincerely！

**Comment by zhanglonghao1992** [comment]
> I just upload a toy example of MI minimization under Gaussian setups. Hope it could help you understand our MI minimization algorithm. For your questions:
> 
> 1. Yes, the structure of q(y|x) influence the performance significantly. The estimators in this repo are under the simplest cases. You may change the architecture of q(y|x) to fit your project.
> 2. The training of MI minimization is somehow similar to the training of GAN. I think you can try different learning rates and different structures of q(y|x) for the MI estimator. If the estimation is negative, that is because your MI estimator is not well learned. You need to update the parameters of the MI estimator (by maximizing the loglikelihood) more frequently or enlarge the learning rate.

Hi, can I ONLY train MI estimator for some epochs, then starting to train MI minimization?   @Linear95

**Comment by Linear95** [comment]
> > I just upload a toy example of MI minimization under Gaussian setups. Hope it could help you understand our MI minimization algorithm. For your questions:
> > 
> > 1. Yes, the structure of q(y|x) influence the performance significantly. The estimators in this repo are under the simplest cases. You may change the architecture of q(y|x) to fit your project.
> > 2. The training of MI minimization is somehow similar to the training of GAN. I think you can try different learning rates and different structures of q(y|x) for the MI estimator. If the estimation is negative, that is because your MI estimator is not well learned. You need to update the parameters of the MI estimator (by maximizing the loglikelihood) more frequently or enlarge the learning rate.
> 
> Hi, can I ONLY train MI estimator for some epochs, then starting to train MI minimization? @Linear95

You can't. Because once the MI minimization starts, the joint distribution of (X,Y) begins changing. Then the MI estimator loses its estimation accuracy. Therefore, the MI estimator requires continuous updates during the minimization process.

**Comment by zhanglonghao1992** [comment]
> > > I just upload a toy example of MI minimization under Gaussian setups. Hope it could help you understand our MI minimization algorithm. For your questions:
> > > 
> > > 1. Yes, the structure of q(y|x) influence the performance significantly. The estimators in this repo are under the simplest cases. You may change the architecture of q(y|x) to fit your project.
> > > 2. The training of MI minimization is somehow similar to the training of GAN. I think you can try different learning rates and different structures of q(y|x) for the MI estimator. If the estimation is negative, that is because your MI estimator is not well learned. You need to update the parameters of the MI estimator (by maximizing the loglikelihood) more frequently or enlarge the learning rate.
> > 
> > 
> > Hi, can I ONLY train MI estimator for some epochs, then starting to train MI minimization? @Linear95
> 
> You can't. Because once the MI minimization starts, the joint distribution of (X,Y) begins changing. Then the MI estimator loses its estimation accuracy. Therefore, the MI estimator requires continuous updates during the minimization process.

In each MI minimization iter, can I train and optimize the MI estimator for several iters with the same batch of data?

**Comment by Linear95** [comment]
> > > > I just upload a toy example of MI minimization under Gaussian setups. Hope it could help you understand our MI minimization algorithm. For your questions:
> > > > 
> > > > 1. Yes, the structure of q(y|x) influence the performance significantly. The estimators in this repo are under the simplest cases. You may change the architecture of q(y|x) to fit your project.
> > > > 2. The training of MI minimization is somehow similar to the training of GAN. I think you can try different learning rates and different structures of q(y|x) for the MI estimator. If the estimation is negative, that is because your MI estimator is not well learned. You need to update the parameters of the MI estimator (by maximizing the loglikelihood) more frequently or enlarge the learning rate.
> > > 
> > > 
> > > Hi, can I ONLY train MI estimator for some epochs, then starting to train MI minimization? @Linear95
> > 
> > 
> > You can't. Because once the MI minimization starts, the joint distribution of (X,Y) begins changing. Then the MI estimator loses its estimation accuracy. Therefore, the MI estimator requires continuous updates during the minimization process.
> 
> In each MI minimization iter, can I train and optimize the MI estimator for several iters with the same batch of data?

Yes, I think this could work.
